mlir/lib/Driver/CompilerDriver.cpp

```diff
@@ -350,10 +464,12 @@ LogicalResult runLowering(const CompilerOptions &options, MLIRContext *ctx, Modu
     // For each pipeline-terminating pass, print the IR into the corresponding dump file and
     // into a diagnostic output buffer. Note that one pass can terminate multiple pipelines.
     auto afterPassCallback = [&](Pass *pass, Operation *op) {
```

Comment:
It would be nice to use the [runBeforePipeline](https://mlir.llvm.org/doxygen/classmlir_1_1PassInstrumentation.html#ae255ba9d617b707e34be7616e0358869) and [runAfterPipeline](https://mlir.llvm.org/doxygen/classmlir_1_1PassInstrumentation.html#a195e695bcacbcae0ba69d95dae23d383) instrumentation at the pipeline level, instead of the individual pass instrumentation. With that we can hopefully instrument individual pipelines, and also prevent the program size measurement from being included in the timing measurement.

Looking at this little snippet:
```
    - capture:
        walltime: 5.4755
        cputime: 5.289
        - parseMLIRSource
          walltime: 30.0263
          cputime: 0
        - parseMLIRSource
          program size: 59
        - 0_canonicalize
          program size: 47
        - runMLIRPasses
          walltime: 6.204
          cputime: 0
```
we would ideally prevent:
- the step name to be printed twice when both timing and program size is requested
- the timing steps should be a superset of the program size steps, that is some steps may have no program size, but there shouldn't be a step that has a program size but no timings